### ISSUE TITLE
Add a tag for Furigana in RichTextLabel

### ISSUE THREAD

**Issue by lpares12** [godotengine/godot-proposals#5876]
### Describe the project you are working on

Developing a Japanese learning app/game.

### Describe the problem or limitation you are having in your project

The Japanese language relies on this thing called Furigana, which basically is a text on top of Kanji to display its pronunciation. You can see an example [here](https://japanbase.net/upload/content/furigana.png). As you can see in the image, a Kanji 私 has the text わたし on top, which defines the reading of the Kanji.

This is used in all reading materials (books, newspapers, games,...) broadly, meaning any app or game that uses the Japanese language will be using Furigana at some point, and currently there is no easy way to implemt this in Godot. You could technically overlap labels on top of a RichTextLabel containing Kanji. But then you would need to manually modify line heights of the lines in the RichTextLabel to accomodate space for the furigana.

There is no good current solution to do this right now. In my case I need to display Furigana so users know how to pronounce the Kanji that they learn in my game.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

My idea is to add a `[furigana]` tag in RichTextLabel's BBCode. The user will be able to add bbcode_text such as:

```
hello, this is a [furigana=brief]example[/furigana]
```

To display something like: `hello, this is a example` with the `brief` word centered on top of the work `example` in a smaller font.

And also a new font entry (furigana_font) will be added to the theme of RichTextLabel, since 99% of the times the font for the Furigana will be smaller than the normal text.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

You can see an example of how this is implemented in: https://github.com/lpares12/godot/commits/feat_furigana which is based on the current (30/11/2022) 3.5 branch of godotengine repo.

And this is how it looks in the editor.

![Screenshot from 2022-11-30 11-27-31](https://user-images.githubusercontent.com/9217676/204809728-7bf04985-6ba1-497d-95b6-94d3f7cc1caa.png)

For the bbcode_text `This is a furigana demo [furigana=abc]attached[/furigana]. You can see abc on top of the word "attached"`

![Screenshot from 2022-11-30 11-29-25](https://user-images.githubusercontent.com/9217676/204809730-860cc51e-1fd5-4ea4-8c6a-12ef9ffd241b.png)

For now this implementation does not support effects and so on, and whenever the Furigana is bigger than the text, it will move out of the container and be "cut".

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, this is not something easy to workaround. And I can't think of any way to do this that does not take a lot of hours to implement.

### Is there a reason why this should be core and not an add-on in the asset library?

Because this will open up a lot of possiblities for Japanese games and apps to be developed with Godot. As I said before, this is a must-have tool for Japanese games. Mostly all books have this, even for Japanese native speakers, so this is not only used for japanese learning purposes.

I'm not sure this can be easily integrated as an add-on or plugin, since it is necessary to change the line-height of individual lines within a RichTextLabel and break lines when the word does not fit in the line. Maybe someone with a bit more experience could do this, but for me it was easier to just implement it into the code.

**Comment by lpares12** [comment]
By the way, I spoke with someone in the godot contributors chat and I was told new features should be added to Godot 4.x and be backported to 3.x, but unless this is gonna be integrated I prefer not to spend time doing it for Godot 4.x (since I currently don't use it). That's why for now the feature I made is base on 3.5 branch.

**Comment by YuriSizov** [comment]
> By the way, I spoke with someone in the godot contributors chat and I was told new features should be added to Godot 4.x and be backported to 3.x, but unless this is gonna be integrated I prefer not to spend time doing it for Godot 4.x (since I currently don't use it). That's why for now the feature I made is base on 3.5 branch.

Without an implementation for 4.x no new feature will be merged. We don't want to create discrepancy between versions such that different branches start to diverge from each other. Also keep in mind that implementing this for 4.x and for 3.x would be very different.

**Comment by lpares12** [comment]
You are right, there are some resemblances but it is true that some things have changed a lot. That's why I do not plan to do it for Godot 4 for now, if it's not gonna be added, since I'm using Godot 3 and I can use my fork for that. But still other people will benefit from this addition.

**Comment by Psychpsyo** [comment]
I'd suggest using [ruby][/ruby] as the bbcode for this instead of [furigana][/furigana].
Ruby text is the more general term for this concept when applied to more than just Japanese, like Korean or Chinese.

Additionally it'd be the same terminology that HTML uses for this with its <ruby> tag. (although their version is a bit more complicated)

**Comment by lpares12** [comment]
Changing it to ruby sounds good to me. I did not know about it, but if it's a general term for other languages too, makes total sense.

**Comment by BLumia** [comment]
Linking https://github.com/godotengine/godot-proposals/issues/3417 here since it's related. Also +1 for using `[ruby]` as the tag name (maybe the proposal can also be renamed to `Add a [ruby] tag for Furigana in RichTextLabel` as well).

**Comment by neono-simneo** [comment]
A [ruby] tag would be a great addition and help out with CJK localization immensely. I hope someone with more skills than me can add this to the engine.

**Comment by Zemeio** [comment]
Hey guys, I started working on this for latest version of Godot 4, but I have my questions about if this is the best approach for a ruby text. The problem I see is that some times we want to apply color to only one word, and it comes at a different place in the base text.

E.g., if I have something like: `Woman in [color=red]red[/color]` and I want this to be the ruby text above `[color=red]赤い[/color]ドレスの女`, this would not be possible because we cannot put tags in the ruby text. Is this an acceptable limitation?

**Comment by neono-simneo** [comment]
> E.g., if I have something like: `Woman in [color=red]red[/color]` and I want this to be the ruby text above `[color=red]赤い[/color]ドレスの女`, this would not be possible because we cannot put tags in the ruby text. Is this an acceptable limitation?

Perfectly acceptable, especially since normal usage will only been adding ruby to a single word, eg: writing "あか" above "赤", so something like: 

[color=red][ruby=あか]赤[/ruby]い[/color]ドレスの女

It could possibly inherit the colour from the outer tag?

Another example:

[color=red][ruby=あか]赤[/ruby]い[/color]ドレスの[ruby=女]レイディー[/ruby]

So here, 赤い will be red, あか will be above 赤, and 女 will be above レイディー, but it would be the default text colour.

(In this example, the first ruby is adding a phonetic guide to a Chinese character, and the second ruby is adding a translation to an english word, both common usages of furigana)

Thanks so much for your hard work!

**Comment by Zemeio** [comment]
Thank you for the response!

I think I may not have explained very well. For translation I think that's fine, but for ateji (where the Japanese use it for style) it becomes a little bit harder.

Here is something that would not be possible with the current approach:

<img width="2172" height="724" alt="Image" src="https://github.com/user-attachments/assets/ae217b9f-7dc6-4a39-9320-f060d78c4468" />

I put both of them in Japanese to show, but I think it maybe the case that this would be better handled by making a future modification.

With my current implementation here is how it looks with your examples.

<img width="427" height="180" alt="Image" src="https://github.com/user-attachments/assets/bc30d22f-fcc8-4cf5-ab7f-3333a51c6cb5" />

I think even though the example I shared would not be possible, the implementation is good enough for this ticket, so after I am done reading through the guidelines and documentations I think I will do the PR.
